### PR TITLE
IGNITE-10306 SQL: Transform subqueries to JOINs when possible

### PR DESCRIPTION
**Pull request by korlov42**
This changes cover follow cases:

- rewrite subqueries from select list
- rewrite subqueries from table list
- rewrite subqueries in IN and EXISTS expression

For now only one-level subqueries are supported.

Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridQueryOptimizer.java [hunk 1]
```diff
@@ -0,0 +1,865 @@
...
+                i--;
+
+            wasPulledOut |= locPulledOut;
+        }
+
+        // pull out sub-queries from table list
+        do {
+            locPulledOut = false;
+
+            if (ELEMENT_WITH_ALIAS_WITH_SUBQUERY.test(null, select.from()))
```
Comment (tledkov-gridgain) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridQueryOptimizer.java:166]: Does it make sense to check this condition at each iteration (and the same condition at the loops below)?
Please clarify. It's not clear to me.

Comment (korlov42) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridQueryOptimizer.java:166]: No, it doesn't. Good catch!

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridQueryOptimizer.java [hunk 2]
```diff
@@ -0,0 +1,865 @@
...
+                ASTNodeFinder.Result res;
+                while ((res = finder.findNext()) != null)
+                    locPulledOut |= pullOutSubQryFromTableList(select, res.getEl(), res.getIdx());
+            }
+
+            wasPulledOut |= locPulledOut;
+        }
+        while (locPulledOut);
+
+        // pull out sub-queries from table list
```
Comment (korlov42) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridQueryOptimizer.java:184]: this is duplicate of section above. Should be deleted

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/QueryParser.java [hunk 3]
```diff
@@ -553,6 +553,8 @@ else if (!prepared.isQuery()) {
                 GridCacheTwoStepQuery twoStepQry = null;
 
                 if (splitNeeded) {
+                    GridQueryOptimizer.pullOutSubQueries(selectStmt);
```
Comment (Berkof) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/QueryParser.java:556]: GridQueryOptimizer - the name doesn't show its meaning. Is there any plans to do any other query optimization except subquery extraction? If yes - it should be named like GridSubueryJoinOptimizer

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridQueryOptimizer.java [hunk 4]
```diff
@@ -0,0 +1,865 @@
...
+        if (parent instanceof GridSqlUnion) {
+            GridSqlUnion union = (GridSqlUnion)parent;
+
+            wasPulledOut |= pullOutSubQueries(union.left());
+            wasPulledOut |= pullOutSubQueries(union.right());
+
+            return wasPulledOut;
+        }
+
+        // pull out select expressions from SELECT clause
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridQueryOptimizer.java:135]: Think it's good idea, so fixed

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridQueryOptimizer.java (hunk 1 above)
Comment (korlov42) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridQueryOptimizer.java:166]: well, after some investigation i'd decided to get rid of this limitation, so now it does make sense. I left a comment describing the reason in the code

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/QueryParser.java (hunk 3 above)
Comment (korlov42) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/QueryParser.java]: Right now it does only one thing - rewrites subqueries. So ok, let's rename it

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java [hunk 5]
```diff
@@ -0,0 +1,869 @@
...
+        return true;
+    }
+
+    /**
+     * Whether sub-query can be pulled out or not.
+     *
+     * @param subQry Sub-query.
+     * @return Sub-query if it can be pulled out or {@code null}.
+     */
+    private static GridSqlSelect checkSubQueryCanBePulledOut(GridSqlSubquery subQry) {
```
Comment (Berkof) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java:652]: Should be named trySubQueryPullOut or should return boolean/Boolean. Same with JavaDoc - its about possibility of doing something, but actually do it and return the result.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java [hunk 6]
```diff
@@ -0,0 +1,869 @@
...
+
+            if (aggFinder.findNext() != null)
+                return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Check whether table have unique index that can be built with provided column set.
```
Comment (Berkof) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java:306]: table HAS uniq

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java [hunk 7]
```diff
@@ -0,0 +1,869 @@
...
+                    colArr[i] = op.child(i);
+                else if (op.child(i) instanceof GridSqlOperation
+                    && ((GridSqlOperation)op.child(i)).operationType() == NEGATE
+                    && op.child() instanceof GridSqlColumn)
+                    colArr[i] = op.child(i).child();
+            }
+
+            if (colArr[0] == null || colArr[1] == null
+                // reference equality used on purpose here, because we should to get known
+                // whether the column referes to exactly the same table
```
Comment (Berkof) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java:717]: refers

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizerSelfTest.java [hunk 8]
```diff
@@ -0,0 +1,591 @@
...
+import org.apache.ignite.internal.IgniteEx;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ *
+ */
+public class GridSubqueryJoinOptimizerSelfTest extends GridCommonAbstractTest {
```
Comment (Berkof) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizerSelfTest.java:35]: @Test
    public void test222() {
        String outerSqlTemplate = "select e.id, e.name from emp e where exists (%s)";
        String subSql = "select 1 from dep d, dep2 d2 where d.id = d2.id";

        String resSql = String.format(outerSqlTemplate, subSql);

        check(resSql, 2);
    }

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizerSelfTest.java [hunk 9]
```diff
@@ -0,0 +1,591 @@
...
+
+        optimizationEnabled(true);
+
+        List<List<?>> act = cache.query(new SqlFieldsQuery(sql).setEnforceJoinOrder(true)).getAll();
+
+        Assert.assertEquals("Result set mismatch", exp, act);
+
+        String plan = cache.query(new SqlFieldsQuery("explain " + sql)).getAll().get(0).get(0).toString();
+
+        System.out.println(plan);
```
Comment (Berkof) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizerSelfTest.java:544]: println

Comment (korlov42) [reply modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizerSelfTest.java:544]: think it's OK to have println in the tests. It serves debugging purpose.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java [hunk 10]
```diff
@@ -0,0 +1,872 @@
...
+        parent.from(new GridSqlJoin(parentFrom, GridSqlAlias.unwrap(subS.from()), false, null)).child();
+
+        return true;
+    }
+
+    /**
+     * Whether sub-query can be pulled out or not.
+     *
+     * @param subQry Sub-query.
+     * @return Sub-query if it can be pulled out or {@code null}.
```
Comment (Berkof) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java:652]: outdated javadoc

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java [hunk 11]
```diff
@@ -0,0 +1,872 @@
...
+                    sqlCols.add(colArr[0]);
+
+                if (colArr[1] != null && colArr[1].expressionInFrom() == subS.from())
+                    sqlCols.add(colArr[1]);
+            }
+        }
+
+        GridSqlAst subTbl = GridSqlAlias.unwrap(subS.from());
+
+        return subTbl instanceof GridSqlTable && isUniqueIndexExists(sqlCols, (GridSqlTable)subTbl);
```
Comment (Berkof) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java:732]: @Test
    public void test223() {
        String outerSqlTemplate = "select e.id, e.name from emp e where exists (%s) and e.name = ''";
        String subSql = "select 1 from dep d, dep2 d2 where d.id = d2.id and d.id = d2.id2 and d.id = e.id";
        subSql = "select 1 from dep2 d2 where (d2.name = '' and d2.id = e.id) or (d2.name = '' and d2.id2 = e.id)";
        String resSql = String.format(outerSqlTemplate, subSql);
        check(resSql, 2);
    }

**Review by Berkof — APPROVED** [review]
LGTM

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizerSelfTest.java (hunk 8 above)
Comment (korlov42) [reply modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizerSelfTest.java:35]: testOptimizationShouldNotBeApplied4

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java (hunk 10 above)
Comment (korlov42) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java:652]: fixed

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java (hunk 11 above)
Comment (korlov42) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java:732]: testOptimizationShouldNotBeApplied2

**Pull request closed without merging**